build.gradle.kts

```diff
@@ -17,9 +17,24 @@ java {
     withSourcesJar()
 }
 
+modularity {
+    mixedJavaRelease(8)
```

Comment:
Do I understand correctly that this means there is only one compilation and it now targets bytecode level 8 (yet still compiles and adds `module-info.java` to the produced jar)?

Reply:
> Do I understand correctly that this means there is only one compilation and it now targets bytecode level 8 (yet still compiles and adds `module-info.java` to the produced jar)?

This is a feature of [gradle-modules-plugin](https://github.com/java9-modularity/gradle-modules-plugin).

After setting it, it will compile the project with Java 8 as the target, and then separately compile module-info.java with Java 9 as the target.

This plugin is implicitly introduced by the JavaFX Gradle plugin, so I can use it directly. 

I updated this PR to declare an explicit dependency on the plugin.

Reply:
Ok this is what I understood, then. I'm quite reluctant to make the change this way. This means we won't be able to compile with higher level bytecode for as long as we want to support Java 8. A better way IMO to support Java 8 would be to publish a separate variant of the library (as a different artifact), which would also avoid to make compromises on transitive dependencies (which sooner or later can become a problem). This is more complicated to setup, and it's the reason why I had not done it when initially moving to the module system, but it seems more correct.

Reply:
> Ok this is what I understood, then. I'm quite reluctant to make the change this way. This means we won't be able to compile with higher level bytecode for as long as we want to support Java 8. A better way IMO to support Java 8 would be to publish a separate variant of the library (as a different artifact), which would also avoid to make compromises on transitive dependencies (which sooner or later can become a problem). This is more complicated to setup, and it's the reason why I had not done it when initially moving to the module system, but it seems more correct.

I can't understand the meaning of compile with higher level bytecode for FX Gson. Could you tell me why?

Reply:
There is no specific feature requiring higher bytecode for FX Gson specifically right now, but going back to level 8 and being stuck there "forever" is what makes me reluctant.

Also, creating a single hybrid artifact has other problems. We're effectively lying on dependencies in the pom of such artifact. For Java 9+ projects, this hybrid jar has no transitive dependency on the JavaFX module in its pom.xml (yet needs it) and its bytecode level is 8. And on the other hand, for Java 8 projects, the jar contains a weird `module-info.class` (with bytecode level 9 that it's not able to read) and a `compileOnly` dependency on the Java FX module that it doesn't need.

Reply:
> There is no specific feature requiring higher bytecode for FX Gson specifically right now, but going back to level 8 and being stuck there "forever" is what makes me reluctant.
> 
> Also, creating a single hybrid artifact has other problems. We're effectively lying on dependencies in the pom of such artifact. For Java 9+ projects, this hybrid jar has no transitive dependency on the JavaFX module in its pom.xml (yet needs it) and its bytecode level is 8. And on the other hand, for Java 8 projects, the jar contains a weird `module-info.class` (with bytecode level 9 that it's not able to read) and a `compileOnly` dependency on the Java FX module that it doesn't need.

In the Java projects I have seen, I have never encountered the need to release cans separately for Java 8.

Even if we have to use the API of a higher version of JDK, we can also use the multi-release jar to merge it into the same jar.

However, declaring the transitive dependency on JavaFX in pom is a disaster, which should not happen.

JavaFX is not a Java library distributed in the traditional way. It involves different platforms (operating systems, architectures) and different distribution methods (using Maven Central, bundled in JDK, or use jmod files).

Declaring a dependency on JavaFX in POM is a very bad behavior, which will only cause problems for users.
The correct way is not to declare the dependence on JavaFX in pom, and the application developer should solve the problem of JavaFX by himself.

Reply:
> Even if we have to use the API of a higher version of JDK, we can also use the multi-release jar to merge it into the same jar.

Multi-release jars are yet another beast, which is [not unanimously considered a good solution to this problem](https://blog.gradle.org/mrjars).

> However, declaring the transitive dependency on JavaFX in pom is a disaster, which should not happen. [...]

Right, maybe in the specific case of JavaFX it's not actually that much of a problem. Though it was just a poor example for a more general concept.

In any case, let's say it's not that big of a problem at the moment. Gson itself is outdated anyway, so I guess it's not a big problem to lock this library into old times as well. There is still the issue of the classpath mode, though.